clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerInterceptors.java

```diff
@@ -82,11 +83,12 @@ public ProducerRecord<K, V> onSend(ProducerRecord<K, V> record) {
      * @param metadata The metadata for the record that was sent (i.e. the partition and offset).
      *                 If an error occurred, metadata will only contain valid topic and maybe partition.
      * @param exception The exception thrown during processing of this record. Null if no error occurred.
+     * @param headers The headers for the record that was sent
      */
-    public void onAcknowledgement(RecordMetadata metadata, Exception exception) {
+    public void onAcknowledgement(RecordMetadata metadata, Exception exception, Headers headers) {
         for (ProducerInterceptor<K, V> interceptor : this.interceptors) {
             try {
-                interceptor.onAcknowledgement(metadata, exception);
+                interceptor.onAcknowledgement(metadata, exception, headers);
```

Comment:
I know this came up during the KIP review, but really the headers ought to be read-only here. You could make a read-only implementation of `Headers` that delegates all read operations to the actual `Headers`. Otherwise, some cunning person might start modifying the headers structure and passing it between interceptors.

Reply:
Ah, no. Not necessary. The producer has already made them read-only. :)

Reply:
Actually, when send successfully, headers are read only. But when send failure, they are NOT. I think it is reasonable to be mutable when send failure, what do you think ❓ 

I have a test case for it now, 988c09f481

Reply:
I think they should be immutable for send failures too. I can't think of a situation where mutating the headers in the interceptor should work so best to prevent it outright.

Reply:
fair enough. I just add `testHeadersFailure` to make sure onAcknowledgement headers is read only
7bd95cf30736a13ac74f6ac2633824b29e49ad33